src/Microsoft.OData.Core/UriParser/Resolver/CaseInsensitiveSchemaElementsCache.cs

```diff
@@ -0,0 +1,121 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="CaseInsensitiveSchemaElementsCache.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.OData.Edm.Vocabularies;
+
+namespace Microsoft.OData.Edm
+{
+    /// <summary>
+    /// Cache used to store schema elements using case-normalized names
+    /// to speed up case-insensitive model lookups. The cache is populated
+    /// up front so that if an item is not found in the cache, we can assume
+    /// it doesn't exist in the model. For this reason, it's important that
+    /// the model be immutable.
+    /// </summary>
+    internal sealed class CaseInsensitiveSchemaElementsCache
+    {
+        // We create different caches for different types of schema elements because all current usage request schema elements
+        // of specific types. If we were to use a single dictionary <string, ISchemaElement> we would need
+        // to do additional work (and allocations) during lookups to filter the results to the susbset that matches the request type.
+        private readonly Dictionary<string, List<IEdmSchemaType>> schemaTypesCache = new Dictionary<string, List<IEdmSchemaType>>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, List<IEdmOperation>> operationsCache = new Dictionary<string, List<IEdmOperation>>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, List<IEdmTerm>> termsCache = new Dictionary<string, List<IEdmTerm>>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Builds a case-insensitive cache of schema elements from
+        /// the specified <paramref name="model"/>.
+        /// </summary>
+        /// <param name="model">The model whose schema elements to cache. This model should be immutable. See <see cref="ExtensionMethods.MarkAsImmutable(IEdmModel)"/>.</param>
+        public CaseInsensitiveSchemaElementsCache(IEdmModel model)
+        {

```

Comment:
@habbes Should we do a null check for the `model` here? In addition, would this be a good place to check if the model is immutable or would that contribute to unnecessary coupling of things that address separate concerns?

Reply:
I already check whether the model is immutable in the caller, before calling in the constructor. So I reasoned it would be of use to call it again in the constructor because this is an internal class, it would add error-handling complexity and I didn't want to call `IsImmutable()` twice in the same code path because there's some cost associated with looking up the model's annotations (which is where we store whether or not it's immutable)

Reply:
I think we could do a `Debug.Assert`. This is an internal class that shouldn't be called if the model is null. If it does get called with null, then something else must have gone wrong either in the ODataUriResolver or higher up the chain.